src/NMonad/Core.hs

```diff
@@ -47,25 +46,32 @@ data NEnv = NEnv
 data NConfig = NConfig
   { defaultTimeout :: Int      -- ^ Default timeout for notification popups.
   , disableReplacement :: Bool -- ^ Whether to disable replacing notifications.
+
+  , dbusNotificationHook :: DBusNotification -> N (Maybe DBusNotification)
+    -- ^ Process notifications as soon as they are received from DBus, optionally discarding it.
+  , notificationHook :: Notification -> N (Maybe Notification)
+    -- ^ Process notifications after processed by nmonad, optionally discarding it.
   }
 
 instance Default NConfig where
   def = NConfig
     { defaultTimeout = 3
     , disableReplacement = False
+    , dbusNotificationHook = return . Just
+    , notificationHook = return . Just
     }
 
 -- | The mutable state of the daemon.
 data NState = NState
-  { notificationCount :: Int        -- ^ Counter for notification IDs.
-  , notifications :: [Notification] -- ^ List of received notifications.
-  } deriving (Show)
+  { notificationCount :: Word32              -- ^ Counter for notification IDs.
+  , notifications :: Map Word32 Notification -- ^ Notifications indexed by ID.
```

Comment:
What was the motivation behind changing data types here to Word32? Is it related to the use of Lens or is the choice due to another reason?

Reply:
There were two: this state will (should) not be directly accessed and manipulated by an end user, so there's no advantage to using `Int` (implicitly imported from `Prelude`) over `Word32`, which is already the type used to represent notification IDs; additionally, `Int` is a signed integer, meaning `generate arbitrary` (which we don't use anymore for this record) half the time would yield negative notification counts. `Word32` is an unsigned integer type, it doesn't have that problem.

Reply:
Makes sense!